cmd/plugin/secret_generation.go

```diff
@@ -0,0 +1,454 @@
+		panic(err)
+	}
+
+	rootCmd.Flags().StringVar(&name, "name", "", "name for the secret (defaults to context name)")
+	rootCmd.Flags().StringVarP(&namespace, "namespace", "n", "dns-operator-system", "namespace to create the secret in")
+	rootCmd.Flags().StringVarP(&serviceAccount, "service-account", "a", "dns-operator-remote-cluster", "service account name to use")
+}
+
+func secretGeneration(_ *cobra.Command, _ []string) error {
+	log = logf.Log.WithName("secret-generation")
+
+	dirtyStr := os.Getenv("KUBECTL_DNS_DIRTY")
+	if strings.Compare(strings.ToLower(dirtyStr), "true") == 0 {
+		log.V(1).Info("kubectl-dns running in dirty mode")
+		dirty = true
+	}
+
+	log.V(1).Info("Set the secret name to context name if not set", "context", contextName, "name", name)
+	if len(name) == 0 {
+		name = contextName
+	}
+
+	err := checkKubectl(log)
+	if err != nil {
+		return err
+	}
+
+	clusterCA, err := getClusterCA(log)
+	if err != nil {
+		return err
+	}
+
+	clusterServer, err := getClusterServer(log)
+	if err != nil {
+		return err
+	}
+
+	serviceAccountToken, err := generateServiceAccountToken(log)
+	if err != nil {
+		return err
+	}
+
+	log.V(1).Info("creation directory to store temporary files")
+	tmpDir, err := os.MkdirTemp("", "kudectl-dev")
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := tidyFiles(log, tmpDir); err != nil {
+			log.Error(err, "deferred call failed")
+		}
+	}()
+
+	kubeConfig := defineKubeConfig(log, clusterServer, clusterCA, serviceAccountToken)
+	kubeConfigFile, err := saveKubeConfig(log, kubeConfig, tmpDir)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := tidyFiles(log, kubeConfigFile.Name()); err != nil {
+			log.Error(err, "deferred call failed")
+		}
+	}()
+
+	err = verifyKubeConfigConnection(log, kubeConfigFile.Name())
+	if err != nil {
+		return err
+	}
+
+	secret, err := generateSecret(log, kubeConfig)
+	if err != nil {
+		return err
+	}
+
+	secretFile, err := saveSecret(log, *secret, tmpDir)
+	if err != nil {
+		return err
+	}
+	defer func() {
+		if err := tidyFiles(log, secretFile.Name()); err != nil {
+			log.Error(err, "deferred call failed")
+		}
+	}()
+
+	err = applySecretToCluser(log, secretFile.Name())
+	if err != nil {
+		return err
+	}
+
+	log.Info("The operator should now be able to discover, and connect to this cluster")
+
+	return nil
+}
+
+func getClusterCA(log logr.Logger) (string, error) {
+	log.V(1).Info("Get the cluster CA certificate from the remote cluster")
+	args := []string{
+		fmt.Sprintf("--context=%v", contextName),
+		"config",
+		"view",
+		"--raw",
+		"--minify",
+		"--flatten",
+		"-o",
+		"jsonpath='{.clusters[].cluster.certificate-authority-data}'",
+	}
+	clusterCARaw, err := exec.Command(kubectl, args...).Output()
+	if err != nil {
+		if log.V(1).Enabled() {
+			log.V(1).Error(err, "unable to get certificate authority from kubeconfig", "context", contextName)
+		}
+		return "", fmt.Errorf("unable to get certificate for %v from kubeconfig", contextName)
+	}
```

Comment:
As discussed, we should update all these "kubectl" commands so that we can properly capture and output any errors.